src/jit/codegenxarch.cpp

```diff
@@ -2689,84 +2632,97 @@ CodeGen::genStoreRegisterReturnInLclVar(GenTreePtr treeNode)
+                                        regNumber* retRegPtr1)
+{
+    *retRegPtr0 = REG_NA;
+    *retRegPtr1 = REG_NA;
+
+    bool firstIntUsed = false;
+    bool firstFloatUsed = false;
+
+    if (type0 != TYP_UNKNOWN)
+    {
+        if (varTypeIsI(type0) || varTypeIsIntegral(type0))
         {
-            if (structDesc.eightByteClassifications[0] == SystemVClassificationTypeIntegerReference ||
-                structDesc.eightByteClassifications[0] == SystemVClassificationTypeInteger)
-            {
-                retReg0 = REG_INTRET;
-                firstIntUsed = true;
-            }
-            else if (structDesc.eightByteClassifications[0] == SystemVClassificationTypeSSE)
+            *retRegPtr0 = REG_INTRET;
+            firstIntUsed = true;
+        }
+        else if (varTypeIsFloating(type0))
+        {
+            *retRegPtr0 = REG_FLOATRET;
+            firstFloatUsed = true;
+        }
+        else
+        {
+            assert(false && "Invalid eightbyte type");
+        }
+    }
+
+    if (type1 != TYP_UNKNOWN)
+    {
+        if (varTypeIsI(type1) || varTypeIsIntegral(type1))
+        {
+            if (firstIntUsed)
             {
-                retReg0 = REG_FLOATRET;
-                firstFloatUsed = true;
+                *retRegPtr1 = REG_INTRET_1;
             }
             else
             {
-                assert(false && "Invalid eightbyte type");
+                *retRegPtr1 = REG_INTRET;
             }
         }
-
-        if (type1 != TYP_UNKNOWN)
+        else if (varTypeIsFloating(type1))
         {
-            if (structDesc.eightByteClassifications[1] == SystemVClassificationTypeIntegerReference ||
-                structDesc.eightByteClassifications[1] == SystemVClassificationTypeInteger)
+            if (firstFloatUsed)
             {
-                if (firstIntUsed)
-                {
-                    retReg1 = REG_INTRET_1;
-                }
-                else
-                {
-                    retReg1 = REG_INTRET;
-                }
-            }
-            else if (structDesc.eightByteClassifications[1] == SystemVClassificationTypeSSE)
-            {
-                if (firstFloatUsed)
-                {
-                    retReg1 = REG_FLOATRET_1;
-                }
-                else
-                {
-                    retReg1 = REG_FLOATRET;
-                }
+                *retRegPtr1 = REG_FLOATRET_1;
             }
             else
             {
-                assert(false && "Invalid eightbyte type");
+                *retRegPtr1 = REG_FLOATRET;
             }
         }
-
-        if (retReg0 != REG_NA)
-        {
-            getEmitter()->emitIns_S_R(ins_Store(type0), size0, retReg0, lclVarPtr->gtLclNum, offset0);
-        }
-
-        if (retReg1 != REG_NA)
+        else
         {
-            getEmitter()->emitIns_S_R(ins_Store(type1), size1, retReg1, lclVarPtr->gtLclNum, offset1);
+            assert(false && "Invalid eightbyte type");
```

Comment:
Should this be a noway_assert or unreached() so that we get to see a failure even in retail instead of silent bad codegen?